### PR TITLE
Feat hierarchical facet root path and show parent level

### PR DESCRIPTION
**Pull request by maxiloc**
I managed !!! :smile: 
cc @vvo

### CODE REVIEW

**Comment by maxiloc** [comment]
@vvo My instantsearch.js PR for those two things are ready and waiting for this one to be merged

**Comment by vvo** [comment]
Will have a look this WE or monday morning, congrats because that's a good Friday night achievement!

File: README.md [hunk 1]
```diff
@@ -490,6 +490,47 @@ The available sort tokens are:
 - name
 - path
 
+##### Restrict results and hierarchical values to non-root level
+
+By default the hierarchical facet will need the root level
```
Comment (vvo) [inline README.md]: I feel the documentation is not easy to understand here.

Maybe we can do an example based on your usecase:

```
let's say you have a lot of levels:
- fruits
  - yellow
    - citrus
      - spicy

But you only want to get the values starting at "citrus", you could do: prefixPath..
```

Something like that?

File: README.md [hunk 2]
```diff
@@ -490,6 +490,47 @@ The available sort tokens are:
...
+  ]
+});
+
+helper.toggleRefine('categories.lvl0', 'fruits');
+
+```
+
+##### Hide parent level of current parent level
+
+By default the hierarchical facet is going to return the last level under the current refinement and his
```
Comment (vvo) [inline README.md]: > "the child and parent facet values"

File: README.md [hunk 3]
```diff
@@ -490,6 +490,47 @@ The available sort tokens are:
...
+parent level to have some context.
+
+If you do not want this parent level you can set showParentLevel to false
```
Comment (vvo) [inline README.md]: If you do not want to get the parent level values, you can set..

File: src/SearchResults/generate-hierarchical-tree.js [hunk 4]
```diff
@@ -78,18 +80,29 @@ function generateHierarchicalTree(sortBy, hierarchicalSeparator, hierarchicalPre
   };
 }
 
-function filterFacetValues(parentPath, currentRefinement, hierarchicalSeparator) {
+function filterFacetValues(parentPath, currentRefinement, hierarchicalSeparator, hierarchicalRootPath,
+                           hierarchicalShowParentLevel) {
   return function(facetCount, facetValue) {
-    // we always want root levels
-    return facetValue.indexOf(hierarchicalSeparator) === -1 ||
+    // we want the facetValue to be under the hierarchicalPrefix in the tree
```
Comment (vvo) [inline src/SearchResults/generate-hierarchical-tree.js]: I don't get the comment

File: README.md [hunk 5]
```diff
@@ -490,6 +490,47 @@ The available sort tokens are:
...
+
+You can specify an prefix path to filter the hierarchical values
+
+```
+var helper = algoliasearchHelper(client, indexName, {
+  hierarchicalFacets: [{
+    name: 'products',
+    attributes: ['categories.lvl1'],
+    sortBy: ['count:desc', 'name:asc'] // first show the most common values, then sort by name
+    prefixPath: 'fruits'
```
Comment (vvo) [inline README.md]: So is it rootPath or prefixPath? (it's rootPath :D)

**Comment by maxiloc** [comment]
@vvo Rebased with the modification

**Comment by maxiloc** [comment]
@vvo As we discussed, new usage is now:

```
var helper = algoliasearchHelper(client, indexName, {
  hierarchicalFacets: [{
    name: 'products',
    attributes: ['categories.lvl0', 'categories.lvl1'],
    rootPath: 'fruits'
  }]
});

helper.toggleRefine('products', 'fruits'); // Refine on the rootPath
```

Way better :smile:

File: README.md [hunk 6]
```diff
@@ -490,6 +490,51 @@ The available sort tokens are:
...
+    name: 'products',
+    attributes: ['categories.lvl0', 'categories.lvl1'],
+    rootPath: 'fruits'
+  }]
+});
+
+helper.toggleRefine('products', 'fruits'); // Refine on the rootPath
+```
+
+This setting **need** the state to have the rootPath as a refinement.
```
Comment (bobylito) [inline README.md]: Could it be done internally?

File: src/SearchResults/generate-hierarchical-tree.js [hunk 7]
```diff
@@ -78,18 +86,28 @@ function generateHierarchicalTree(sortBy, hierarchicalSeparator, hierarchicalPre
...
       // keep the facetValue
       facetValue.indexOf(hierarchicalSeparator) === -1 &&
       currentRefinement.indexOf(hierarchicalSeparator) === -1 ||
       // currentRefinement is a child of the facet value
-      currentRefinement.indexOf(facetValue + hierarchicalSeparator) === 0 ||
+      currentRefinement.indexOf(facetValue) === 0 ||
       // facetValue is a child of the current parent, add it
-      facetValue.indexOf(parentPath + hierarchicalSeparator) === 0;
+      facetValue.indexOf(parentPath + hierarchicalSeparator) === 0
+      && (hierarchicalShowParentLevel || facetValue.indexOf(currentRefinement) === 0);
```
Comment (bobylito) [inline src/SearchResults/generate-hierarchical-tree.js]: is the indentation correct here?

File: README.md [hunk 8]
```diff
@@ -490,6 +490,51 @@ The available sort tokens are:
...
+```
+var helper = algoliasearchHelper(client, indexName, {
+  hierarchicalFacets: [{
+    name: 'products',
+    attributes: ['categories.lvl0', 'categories.lvl1', 'categories.lvl2', 'categories.lvl3'],
+    rootPath: 'fruits > yellow > citrus'
+  }]
+});
+
+helper.toggleRefine('products', 'fruits > yellow > citrus'); // Refine on the rootPath
```
Comment (vvo) [inline README.md]: We can do this internally as proposed by @bobylito.

Also, can you add a throw if the user tries to `toggleRefine()` on a path that is not part of the rootpath?

**Comment by maxiloc** [comment]
ready (I hope) !

File: README.md [hunk 9]
```diff
@@ -490,6 +490,49 @@ The available sort tokens are:
...
+Let's say you have a lot of levels:
+
+```
+- fruits
+  - yellow
+    - citrus
+      - spicy
+```
+
+But you only want to get the values starting at "citrus", you can use `rootPath`
```
Comment (vvo) [inline README.md]: The examples is not good, I had fixed it but you rebased and force pushed :D

**Comment by maxiloc** [comment]
ready (again)

File: src/SearchParameters/index.js [hunk 10]
```diff
@@ -1382,10 +1398,22 @@ SearchParameters.prototype = {
...
+    return hierarchicalFacet.rootPath || null;
+  },
+
+  /**
+   * Helper function to check if we show the parent level of the hierarchicalFacet
+   * @param  {object} hierarchicalFacet
+   * @return {string} returns the hierarchicalFacet.showParentLevel or true as default
+   */
+  _getHierarchicalShowParentLevel: function(hierarchicalFacet) {
+    if (hierarchicalFacet.showParentLevel !== undefined) {
```
Comment (bobylito) [inline src/SearchParameters/index.js]: Why not make it the other way around and check if the value is a boolean, otherwise return true?

**Comment by maxiloc** [comment]
@bobylito made the change

**Comment by vvo** [comment]
LGTM

**Comment by bobylito** [comment]
Still reviewing ;)

**Comment by bobylito** [comment]
LGTM :+1: Great @maxiloc

**Pull request merged**